File: mycelium-py/src/mycelium/runner.py
from __future__ import annotations

import json
import os
import sys
import time
from dataclasses import dataclass
from typing import Any
from urllib.parse import urlparse

from .mock import MOCK_ENABLED, get_mock_response
from .prime import build_goal, prime
from .recorder import record
from .types import RunOutcome, RunResult


@dataclass
class _TinyFishResult:
    success: bool
    data: Any
    raw: str
    steps: list[str]
    errors: list[str]


def _extract_domain(url: str) -> str:
    full = url if url.startswith("http") else f"https://{url}"
    try:
        host = urlparse(full).hostname or url
    except ValueError:
        host = url
    if host.startswith("www."):
        host = host[4:]
    return host


def _call_tinyfish(url: str, goal: str, silent: bool) -> _TinyFishResult:
    try:
        import httpx
    except ImportError as e:
        raise RuntimeError("httpx package not installed; run `pip install httpx`") from e

    api_key = os.environ.get("TINYFISH_API_KEY")
    if not api_key:
        raise RuntimeError("TINYFISH_API_KEY is not set")

    full_url = url if url.startswith("http") else f"https://{url}"
    body = {"url": full_url, "goal": goal}
    headers = {"Content-Type": "application/json", "X-API-Key": api_key}

    steps: list[str] = []
    errors: list[str] = []
    raw_parts: list[str] = []
    data: Any = None
    success = False

    with httpx.Client(timeout=None) as client:
        with client.stream(
            "POST",
            "https://agent.tinyfish.ai/v1/automation/run-sse",
            headers=headers,
            json=body,
        ) as response:
            if response.status_code >= 400:
                text = response.read().decode("utf-8", errors="replace")
                raise RuntimeError(
                    f"TinyFish API error: {response.status_code} {response.reason_phrase}\n{text}"
                )

            for chunk in response.iter_text():
                if not chunk:
                    continue
                raw_parts.append(chunk)
                for line in chunk.split("\n"):
                    if not line.startswith("data: "):
                        continue
                    payload = line[6:]
                    if os.environ.get("MYCELIUM_DEBUG"):
                        print("[SSE RAW]", payload)
                    try:
                        event = json.loads(payload)
                    except json.JSONDecodeError:
                        continue
                    event_type = event.get("type")
                    if event_type == "PROGRESS" and event.get("purpose"):
                        purpose = event["purpose"]
                        steps.append(purpose)
                        if not silent:
                            sys.stdout.write(f"  · {purpose}\n")
                            sys.stdout.flush()
                    elif event_type == "FAILED" and event.get("message"):
                        errors.append(event["message"])
                    elif event_type == "COMPLETE":
                        data = event.get("result", event.get("data"))
                        success = event.get("status") == "COMPLETED"

    return _TinyFishResult(success=success, data=data, raw="".join(raw_parts), steps=steps, errors=errors)


def run(url: str, goal: str, *, silent: bool = False) -> RunResult:
    domain = _extract_domain(url)

    primed = prime(domain, goal)
    if not silent:
        if primed.hints_loaded > 0:
            plural = "s" if primed.hints_loaded > 1 else ""
            print(f"  + {primed.hints_loaded} hint{plural} loaded for {domain}")
        else:
            print(f"  no knowledge found for {domain} — starting fresh")

    enriched = build_goal(goal, primed)
    t0 = time.monotonic()
    if MOCK_ENABLED:
        mock = get_mock_response(domain, goal)
        success, data, raw, steps, errors = mock.success, mock.data, mock.raw, mock.steps, mock.errors
    else:
        tf = _call_tinyfish(url, enriched, silent)
        success, data, raw, steps, errors = tf.success, tf.data, tf.raw, tf.steps, tf.errors
    duration_ms = int((time.monotonic() - t0) * 1000)

    outcome = RunOutcome(
        domain=domain,
        goal=goal,
        success=success,
        steps=steps,
        errors=errors,
        raw=raw,
        duration_ms=duration_ms,
    )
    recorded = record(outcome)
    if not silent:
        if recorded.hints_extracted > 0:
            plural = "s" if recorded.hints_extracted > 1 else ""
            print(
                f"  + {recorded.hints_extracted} new hint{plural} saved "
                f"({recorded.hints_total} total)"
            )
        else:
            print("  no new hints extracted")

    return RunResult(success=success, data=data, primed=primed, recorded=recorded, raw=raw)

File: js/core/runner.ts
import { prime, buildGoal } from "./prime.ts"
import { record } from "./recorder.ts"
import type { RunOutcome } from "../store/types.ts"
import type { PrimeResult } from "./prime.ts"
import type { RecordResult } from "./recorder.ts"

export interface RunOptions {
  url: string
  goal: string
  silent?: boolean  // suppress console output when used as SDK
}

export interface RunResult {
  success: boolean
  data: any
  primed: PrimeResult
  recorded: RecordResult
  raw: string
}

export async function run(options: RunOptions): Promise<RunResult> {
  const { url, goal, silent = false } = options
  const domain = extractDomain(url)

  // Step 1: prime
  const primed = await prime(domain, goal)
  if (!silent && primed.hintsLoaded > 0) {
    console.log(`  + ${primed.hintsLoaded} hint${primed.hintsLoaded > 1 ? "s" : ""} loaded for ${domain}`)
  } else if (!silent) {
    console.log(`  no knowledge found for ${domain} — starting fresh`)
  }

  // Step 2: call TinyFish
  const enrichedGoal = buildGoal(goal, primed)
  const t0 = Date.now()
  const { success, data, raw, steps, errors } = await callTinyFish(url, enrichedGoal, silent)
  const durationMs = Date.now() - t0

  // Step 3: record
  const outcome: RunOutcome = { domain, goal, success, steps, errors, raw, durationMs }
  const recorded = await record(outcome, { hintsUsedIds: primed.hintsUsedIds })
  if (!silent) {
    if (recorded.hintsExtracted > 0) {
      console.log(`  + ${recorded.hintsExtracted} new hint${recorded.hintsExtracted > 1 ? "s" : ""} saved (${recorded.hintsTotal} total)`)
    } else {
      console.log(`  no new hints extracted`)
    }
  }

  return { success, data, primed, recorded, raw }
}

async function callTinyFish(
  url: string,
  goal: string,
  silent: boolean
): Promise<{ success: boolean; data: any; raw: string; steps: string[]; errors: string[] }> {
  const apiKey = process.env.TINYFISH_API_KEY
  if (!apiKey) throw new Error("TINYFISH_API_KEY is not set")

  const response = await fetch("https://agent.tinyfish.ai/v1/automation/run-sse", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-API-Key": apiKey,
    },
    body: JSON.stringify({ url: url.startsWith("http") ? url : `https://${url}`, goal }),
  })

  if (!response.ok) {
    const body = await response.text().catch(() => "")
    throw new Error(`TinyFish API error: ${response.status} ${response.statusText}\n${body}`)
  }

  const steps: string[] = []
  const errors: string[] = []
  let raw = ""
  let data: any = null
  let success = false

  // Parse SSE stream
  const reader = response.body!.getReader()
  const decoder = new TextDecoder()

  while (true) {
    const { done, value } = await reader.read()
    if (done) break

    const chunk = decoder.decode(value)
    raw += chunk

    // Parse SSE events
    const lines = chunk.split("\n")
    for (const line of lines) {
      if (!line.startsWith("data: ")) continue
      if (process.env.MYCELIUM_DEBUG) console.log("[SSE RAW]", line.slice(6))
      try {
        const event = JSON.parse(line.slice(6))
        if (event.type === "PROGRESS" && event.purpose) {
          steps.push(event.purpose)
          if (!silent) process.stdout.write(`  · ${event.purpose}\n`)
        }
        if (event.type === "FAILED" && event.message) {
          errors.push(event.message)
        }
        if (event.type === "COMPLETE") {
          data = event.result ?? event.data
          success = event.status === "COMPLETED"
        }
      } catch {
        // non-JSON SSE line, skip
      }
    }
  }

  return { success, data, raw, steps, errors }
}

function extractDomain(url: string): string {
  try {
    const u = url.startsWith("http") ? new URL(url) : new URL(`https://${url}`)
    return u.hostname.replace(/^www\./, "")
  } catch {
    return url.replace(/^(https?:\/\/)?(www\.)?/, "").split("/")[0]
  }
}
